File: app/src/main/java/com/singword/app/data/local/prefs/SettingsRepository.kt
package com.singword.app.data.local.prefs

import com.singword.app.data.local.wordbook.WordbookId
import com.singword.app.ui.theme.AppThemeMode

class SettingsRepository(
    private val prefsManager: PrefsManager
) {
    fun getSelectionMap(): Map<WordbookId, Boolean> = prefsManager.getSelectionMap()

    fun getEnabledWordbooks(): List<WordbookId> = prefsManager.getEnabledWordbooks()

    fun getThemeMode(): AppThemeMode = prefsManager.getThemeMode()

    fun setThemeMode(mode: AppThemeMode) {
        prefsManager.setThemeMode(mode)
    }

    fun toggleWordbook(id: WordbookId): ToggleResult {
        val current = prefsManager.isEnabled(id)
        if (current && prefsManager.getEnabledWordbooks().size == 1) {
            return ToggleResult.Rejected("至少保留一个词表开启")
        }
        prefsManager.setEnabled(id, !current)
        return ToggleResult.Accepted
    }
}

sealed interface ToggleResult {
    data object Accepted : ToggleResult
    data class Rejected(val reason: String) : ToggleResult
}

File: ios/singword/Data/Settings/SettingsRepository.swift
import Foundation

enum ToggleResult {
    case accepted
    case rejected(reason: String)
}

final class SettingsRepository {
    private let defaults: UserDefaults
    private let themeModeKey = "theme_mode"

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    func getSelectionMap() -> [WordbookId: Bool] {
        Dictionary(uniqueKeysWithValues: WordbookId.allCases.map { ($0, isEnabled($0)) })
    }

    func getEnabledWordbooks() -> [WordbookId] {
        WordbookId.allCases.filter { isEnabled($0) }
    }

    func getThemeMode() -> AppThemeMode {
        let raw = defaults.string(forKey: themeModeKey) ?? AppThemeMode.system.rawValue
        return AppThemeMode(rawValue: raw) ?? .system
    }

    func setThemeMode(_ mode: AppThemeMode) {
        defaults.set(mode.rawValue, forKey: themeModeKey)
    }

    func toggleWordbook(_ id: WordbookId) -> ToggleResult {
        let current = isEnabled(id)
        if current && getEnabledWordbooks().count == 1 {
            return .rejected(reason: "至少保留一个词表开启")
        }
        setEnabled(id, enabled: !current)
        return .accepted
    }

    func isEnabled(_ id: WordbookId) -> Bool {
        let key = key(for: id)
        if defaults.object(forKey: key) == nil {
            return id == .cet4
        }
        return defaults.bool(forKey: key)
    }

    private func setEnabled(_ id: WordbookId, enabled: Bool) {
        defaults.set(enabled, forKey: key(for: id))
    }

    private func key(for id: WordbookId) -> String {
        id.rawValue
    }
}
